mbed_targets/mbed_targets.py

```diff
@@ -84,22 +88,71 @@ class DatabaseMode(Enum):
 
     OFFLINE = 0
     ONLINE = 1
+    AUTO = 2
+
+
+def get_target(product_code: str, mode: DatabaseMode) -> MbedTarget:
+    """Get an MbedTarget by its product code.
+
+    Args:
+        product_code: the product code to look up in the database.
+        mode: a DatabaseMode enum field.
+    """
+    if mode == DatabaseMode.OFFLINE:
+        return MbedTargetsOffline().get_target(product_code)
+    if mode == DatabaseMode.ONLINE:
+        return MbedTargetsOnline().get_target(product_code)
+    if mode == DatabaseMode.AUTO:
+        try:
+            return MbedTargetsOffline().get_target(product_code)
+        except UnknownTarget:
+            logger.warning("Could not find the requested target in the offline database. Checking the online database.")
+            return MbedTargetsOnline().get_target(product_code)
+    else:
+        raise UnsupportedMode(f"{mode} is not a supported database mode.")
+
+
+class UnknownTarget(ToolsError):
+    """Requested target was not found."""
+
+
+class UnsupportedMode(ToolsError):
+    """The Database Mode is unsupported."""
 
 
 class MbedTargets(Collection):
-    """Interface to the Online Target Database."""
+    """Base Interface to the Target Database."""
+
+    def __init__(self, target_data: Iterable) -> None:
+        """Initialise with a list of targets.
 
-    _target_data: Any
+        Args:
+            target_data: iterable of target data from a target database source.
+        """
+        self._target_data = list(MbedTarget(t) for t in target_data)
```

Comment:
I wonder it if wouldn't be better to make `MbedTargets` inherit from `collections.abc.Set` (it already implements the interface anyway), since it seems to heavily rely on `set`- like operations. As I understand it:
- you'd get `__sub__` for free
- it'd make `_from_iterable` obsolete, as `set([1]) == set(set([1]))`